Add documentation for Spring Boot integration (#4957)
line/armeria: site/src/pages/docs/advanced-spring-boot-integration.mdx

# Using Spring Boot with Armeria

> Spring framework provides powerful features which are necessary for building a web application, such as dependency injection, data binding, AOP, transaction. Armeria is a lightweight microservices framework, its capabilities is comparable to existing full stack web frameworks. It has Rich support for [Apache Thrift](https://thrift.apache.org/) and [gRPC](https://grpc.io/), including [the fancy web console](https://armeria.dev/docs/server-docservice) that enables you to send Thrift and gRPC requests from a web browser. Armeria supports both HTTP/1 and HTTP/2, and these protocols both support [cleartext](https://simple.wikipedia.org/wiki/Cleartext) and TLS cryptographic protocols. Armeria uses [event loop](https://en.wikipedia.org/wiki/Event_loop) based on [Netty](https://netty.io/index.html) which allows to perform non-blocking I/O operations. Therefore, it shows better performance in certain situations such as media streaming, or large file transfer. If you integrate your Spring application with Armeria, you can serve both legacy Spring service and any service made of Armeria such as gRPC or Thrift in the same port. Also, you can gradually migrate your services starting with those expected to improve performance.

Author: Thanks for review! I will make modifications by dividing the line.

> # Using Spring Boot with Armeria

Reviewer: How about aligning with other titles?
```suggestion
# Using Armeria with Spring Boot
```

Author: <img width="277" alt="스크린샷 2023-07-03 오후 9 08 01" src="https://github.com/line/armeria/assets/24784661/fe766a28-69a0-433e-a53f-6daab8a18ed7">

Thanks! The order of the words in the title is intentional. If the word "Armeria" goes first, the overflowed sentence will be omitted. I think the world "Armeria" is less important because we already knows this document is about Armeria. Therefore, I locate the "Spring Boot" first. Also, I'm thinking about changing the other titles as well. What do you think?

Reviewer 2: Good idea!

Reviewer 2: How about just `Spring Boot integration` then, like other integration pages?

Author: Thanks! I think it's good idea!

Reviewer: That sounds awesome. 😉

[Apache Thrift](https://thrift.apache.org/) and [gRPC](https://grpc.io/), including [the fancy web console](https://armeria.dev/docs/server-docservice)
that enables you to send Thrift and gRPC requests from a web browser. Armeria supports both HTTP/1 and HTTP/2,
and these protocols both support [cleartext](https://simple.wikipedia.org/wiki/Cleartext) and TLS cryptographic protocols.
Armeria uses [event loop](https://en.wikipedia.org/wiki/Event_loop) based on [Netty](https://netty.io/index.html)
which allows to perform non-blocking I/O operations. Therefore, it shows better performance in certain situations such as media streaming,
or large file transfer. If you integrate your Spring application with Armeria, you can serve both legacy Spring service
and any service made of Armeria such as gRPC or Thrift in the same port. Also, you can gradually migrate your services
starting with those expected to improve performance.

Armeria provides a way to integrate legacy Spring Boot applications with just simple code.
Additionally, Armeria supports several internal services for monitoring and management purposes.
You can achieve these convenient properties through bean configuration.

> ## Integrate Legacy Spring Boot with Armeria

Reviewer: What do you think of dividing this section into two?
One for integrating with Spring Boot and another one for Spring MVC?
If so, we can distinguish the usage of each artifact and explain it more clearly. e.g.
```java
## Integrating with Spring Boot

<Tip>

Visit [armeria-examples](https://github.com/line/armeria/tree/main/examples/spring-boot-minimal) to
find a fully working example.

</Tip>

First, you need the `armeria-spring-boot3-starter` dependency:
...
// Explain how an Armeria server is configured and started using Spring Boot.
// We can also introduce ArmeriaServerConfigurator.

## Integrating with Spring MVC
// tip and dependency section needs.
// Explain how to integrate Spring MVC with Armeria using TomcatService.

```

Author: I split the section into two parts. However, I placed the "Integrating with Spring MVC" section first, and then the "Integrating with Spring Boot" section. The readers of this document are likely to be divided into three groups: those using Spring with Tomcat, those using Spring WebFlux, and those using Spring as dependency injection tool. I think most people are using Spring with Tomcat

Therefore, I placed the "Integrating with Spring MVC" section at the beginning. Next, I placed the "Integrating with Spring Boot" section for the group using only Spring dependency injection or Boot's auto configuration. The group using Spring WebFlux is likely to refer to other existing documents.

We will plug Armeria into the Spring Boot system. Add the following dependency to your application.

<RequiredDependencies
  dependencies={[
    {
      groupId: 'com.linecorp.armeria',

>       artifactId: 'armeria',

Reviewer: We don't need this dependency because other dependencies have this as API configuration

    },
    {
      groupId: 'com.linecorp.armeria',
      artifactId: 'armeria-spring-boot3-starter',
    },
    {
      groupId: 'com.linecorp.armeria',
      artifactId: 'armeria-tomcat10',
    }
  ]}
/>

> Serve the embedded tomcat service via Armeria using <type://ArmeriaServiceConfigurator> bean.

Reviewer: ArmeriaServiceConfigurator -> ArmeriaServerConfigurator

   };
}
```

## Internal Services

Armeria supports four internal services which are helpful for monitoring and management.
You can use those services by include the service ids to application.yml (or application.properties).
Service ids are `docs`, `health`, `metrics`, `actuator` or you can use them `all`.

```yaml filename=application.yml
armeria:
  ports:
    - port: 8080
      protocols:
        - HTTP
  # Add the below configuration
  internal-services:
    include: docs, health, metrics, actuator  # you can add all of the services using `all`
    port: 8090
```

The `armeria.internal-services.port` is unnecessary. If the port is not specified or left as 0,
it will bind the internal service to a random unused port.
It is ok to set the port of internal service same as one of `armeria.ports`.

<Tip>

  The internal services described below must be configured through the beans described in the documentation.
  Do not use the method of <type://ServerBuilder> directly using <type://ArmeriaServerConfigurator> or
  `Consumer<ServerBuilder>` bean. They may be overwritten by the automatic configuration of Armeria.

</Tip>

> <Tip>
> 
>   If you have configured the server as described in the preceding section of the document,
>   it will also serve on the internal service port. If you want to separate port,
>   use <type://ServerBuilder#defaultVirtualHost()> or <type://ServerBuilder#virtualHost(int)>.
> 
> ```java
> @Bean
> public ArmeriaServerConfigurator armeriaServerConfigurator(TomcatService tomcatService) {
>     return sb -> sb
>             .defaultVirtualHost()  // Add this method.
>             .serviceUnder("/", tomcatService)
>             .and();
> }
> ```
> 
> </Tip>

Author: ```suggestion
```
I remove this tip because the document service can only create documents for the virtual host it belongs to. Therefore, if user applies above tip, user may experience difficulties when using the document service. Also, It seems that the situation mentioned in the tip is not important use case.

# Spring Boot integration

The [Spring](https://spring.io/) framework provides powerful features necessary for building web applications
such as dependency injection, data binding, AOP and transaction management.
Armeria is a lightweight microservices framework that offers capabilities comparable to existing full-stack web frameworks.
It has rich support for [Apache Thrift](https://thrift.apache.org/) and [gRPC](https://grpc.io/).
It also includes a user-friendly [web console](https://armeria.dev/docs/server-docservice)
that allows you to send Thrift and gRPC requests directly from a web browser. Armeria supports both HTTP/1 and HTTP/2,
and these protocols both support [cleartext](https://simple.wikipedia.org/wiki/Cleartext) and TLS cryptographic protocols.
Armeria utilizes an [event loop](https://en.wikipedia.org/wiki/Event_loop) based on [Netty](https://netty.io/index.html),

> enabling non-blocking I/O operations. As a result, it demonstrates better performance in certain scenarios such as streaming media or

Reviewer 3: This description might be useful if this page is published as an independent article of a blog. However, this is a part of Armeria documentation. I don't see we need to repeat Armeria features on this page.

What do you think of dropping from L5 to L11?
```
The [Spring](https://spring.io/) framework provides powerful features necessary for building web applications
such as dependency injection, data binding, AOP and transaction management.
By integrating your Spring application with Armeria, you can serve both legacy ...
```

       container.start();
       return container.getTomcat().getConnector();
   }

   @Bean
   public TomcatService tomcatService(ServletWebServerApplicationContext applicationContext) {
       return TomcatService.of(getConnector(applicationContext));
   }

   @Bean
   public ArmeriaServerConfigurator armeriaServerConfigurator(TomcatService tomcatService) {
       return sb -> sb.serviceUnder("/", tomcatService);
   }
}
```

Add the following properties to your `application.yml` (or `application.properties`) file.
This configuration will prevent the exposure of the embedded Tomcat service and instead expose your Armeria service.
You can explore additional configuration options in <type://ArmeriaSettings>.

```yaml filename=application.yml
# Prevent the embedded Tomcat from opening a TCP/IP port.
server:
  port: -1

armeria:
  ports:
    - port: 8080
      protocols:
        - HTTP
```

With this simple configuration, everything is set up. Now you can serve your legacy Spring service through Armeria.
Although it may not seem like much has changed compared to before, your application is now ready to serve both RPC services
(such as gRPC or Thrift) and RESTful services via Armeria.

> 
> 

Reviewer: ```suggestion

```

## Integrating with Spring Boot

<Tip>

  Visit [armeria-examples/spring-boot-minimal](https://github.com/line/armeria/tree/main/examples/spring-boot-minimal) to
  find a fully working example.

</Tip>

Even if you don't utilize Spring MVC based on web servlets like Tomcat, there is still significant value in combining Spring and Armeria.
You can leverage the automatic configuration and dependency injection features of Spring Boot to develop an Armeria application.

First, you need the `armeria-spring-boot3-starter` dependency:

<RequiredDependencies
  boms={[
    { groupId: 'com.linecorp.armeria', artifactId: 'armeria-bom' }
  ]}
  dependencies={[
    {
      groupId: 'com.linecorp.armeria',
      artifactId: 'armeria-spring-boot3-starter',
    }
  ]}
/>

> Armeria's annotated service is similar to a Spring controller.

Reviewer 3: ```suggestion
Armeria's [annotated service](/docs/server-annotated-service) is similar to a Spring controller.
```

You can declare the annotated service as a bean and inject it into the <type://ServerBuilder>.
Here's an example:

```java filename=TodoAnnotatedService.java
import com.linecorp.armeria.common.HttpResponse;
import com.linecorp.armeria.common.HttpStatus;
import com.linecorp.armeria.server.annotation.*;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

import java.util.List;

// Make this annotated service a bean.

> @Controller

Reviewer: Technically we don't need this annotation for an annotated service. How about removing it?
```suggestion
```

Author: Thank you for confirming. In the conventional Armeria server setup, developers manually create an `AnnotatedService` to run the server. However, in this code example, an `AnnotatedService` instance is obtained using Spring's dependency injection.

In the following code segment, an instance of `TodoAnnotatedService` is injected as a parameter.

```java filename=ArmeriaConfiguration.java
@Bean
public ArmeriaServerConfigurator armeriaServerConfigurator(TodoAnnotatedService todoAnnotatedService) {
   return serverBuilder -> {
       serverBuilder.serviceUnder("/docs", new DocService());
                    .decorator(LoggingService.newDecorator());
                    .annotatedService("/api", todoAnnotatedService);
   };
}
```

Therefore, it's a Spring-managed component, indicated by the addition of the `@Controller` annotation.

Reviewer: Ah, you already left a comment on that. Sorry, I didn't notice it. 😓
But, I'm still worried that this might confuse the user even though we don't need `@Contoller` at all.
What do you think of providing the instance using `@Bean`, then?

Author: Thank! It seems like a much better idea. I will make the changes accordingly. However, the `org.springframework.stereotype.Bean` annotation cannot be applied to classes (types), so I applied the `org.springframework.stereotype.Component` annotation instead. Is that okay?

Author: (Sorry, I misunderstood you mentioned, so I deleted the previous reply I had posted.)

Thank you for checking! Is there a reason to avoid using `@Controller` or `@Component` for injection? When using `@Bean` for dependency injection, we need to add a method like the following:

```java
@Bean
public TodoAnnotatedService todoAnnotatedService(TodoService todoService) {
    return new TodoAnnotatedService(todoService);
}
```

In this case, `TodoAnnotatedService` needs to receive the dependency `TodoService` as a parameter. This is automatically injected using `@Autowired` when using the `@Component` annotation. Therefore, using `@Bean` might lead to longer code.

I prefered to use `@Controller` or `@Component` to demonstrate that using Spring Web MVC with Armeria makes dependency management convenient.

Reviewer: I was expecting to use `@Bean` with a method:
```java
@Bean
Public TodoService todoService() {...}
```
But, it sounds like using `@Component` is much better. 👍

@PathPrefix("/todos")
public class TodoAnnotatedService {

    private final TodoRepository todoRepository;

    // Automatically injected by Spring
    @Autowired
    public TodoAnnotatedService(TodoRepository todoRepository) {
        this.todoRepository = todoRepository;
    }

    @Get("/:id")
    public HttpResponse get(@Param Integer id) {
        Todo todo = todoRepository.get(id);
        if (todo == null) {

>             HttpResponse.of(HttpStatus.NO_CONTENT);

Reviewer: nit:
```suggestion
            return HttpResponse.of(HttpStatus.NO_CONTENT);
```

        }
        return HttpResponse.ofJson(todo);
    }

    @Post
    public HttpResponse create(Todo todo) {
        final int result = todoRepository.create(todo);
        if (result == 0) {
            return HttpResponse.of(HttpStatus.INTERNAL_SERVER_ERROR);
        }
        return HttpResponse.of(HttpStatus.CREATED);
    }
}
```

To inject the annotated service into the <type://ArmeriaServerConfigurator> bean, you can follow this approach:

```java filename=ArmeriaConfiguration.java
@Bean
public ArmeriaServerConfigurator armeriaServerConfigurator(TodoAnnotatedService todoAnnotatedService) {
   return serverBuilder -> {

>        serverBuilder.serviceUnder("/docs", new DocService());
>                     .decorator(LoggingService.newDecorator());
>                     .annotatedService("/api", todoAnnotatedService);

Reviewer: ```suggestion
       serverBuilder.serviceUnder("/docs", new DocService())
                    .decorator(LoggingService.newDecorator())
                    .annotatedService("/api", todoAnnotatedService);
```

   };
}
```

The following configuration prevents the execution of Spring's embedded web server and instead runs Armeria.

```yaml filename=application.yml
# See https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto.webserver.disable.
# The application should not run as a web application and should not start an embedded web server.
spring.main.web-application-type: none

armeria:
  ports:
    - port: 8080
      protocols:
        - http
```

## Internal services

Armeria supports four internal services that are useful for monitoring and management purposes.
You can enable these services by including their service IDs in your `application.yml` (or `application.properties`) file.

> Service ids are `docs`, `health`, `metrics`, `actuator`. You can include one or multiple service IDs based on your requirements.

Reviewer: ```suggestion
Service ids are `docs`, `health`, `metrics`, and `actuator`. You can include one or multiple service IDs based on your requirements.
```

Add the service ID `docs` to `armeria.internal-services.include` configuration.
The `armeria.docs-path` is not necessary, as the default path for the documentation service is `/internal/docs`.

```yaml filename=application.yml
armeria:
  internal-services:
    include: docs
  docs-path: /internal/docs
```

To add custom configuration to your documentation service, you can utilize the <type://DocServiceConfigurator> bean.

```java filename=ArmeriaConfiguration.java
@Bean
public DocServiceConfigurator docServiceConfigurator() {
    return docServiceBuilder -> docServiceBuilder
            .exampleRequests(TodoAnnotatedService.class, "create", "{\"id\":\"42\", \"value\":\"foo bar\"}");
}
```

### Health check

To customize the health check operation, you can utilize the <type://HealthChecker> bean.

Add the service ID `health` to the `armeria.internal-services.include` configuration.
The `armeria.health-check-path` is not necessary, as the default path for the health check service is `/internal/healthcheck`.

```yaml filename=application.yml
armeria:
  internal-services:
    include: health
  health-check-path: /internal/healthcheck
```

> You can create a <type://HealthChecker> bean that implements your custom health check logic,

Reviewer: ```suggestion
You can create a <type://HealthChecker> bean that implements your custom health check logic.
```

> For example, the below code is checking tomcat connector state.

Reviewer: nit: how about:
```suggestion
For example, the below code determines if the server is healthy using Tomcat connector state.
```

```java filename=ArmeriaConfiguration.java
@Bean
public HealthChecker tomcatConnectorHealthChecker(ServletWebServerApplicationContext applicationContext) {
    final Connector connector = getConnector(applicationContext);
    return () -> connector.getState().isAvailable();
}
```

You can also add custom configuration to the service using the <type://HealthCheckServiceConfigurator> bean, similar to the documentation service.

```java filename=ArmeriaConfiguration.java
@Bean
public HealthCheckServiceConfigurator healthCheckServiceConfigurator() {
    return healthCheckServiceBuilder -> healthCheckServiceBuilder
            .updatable(true)
            .startUnhealthy();
}
```

### Collecting metrics

Armeria provides a [built-in metric service](/docs/advanced-metrics) using the [Micrometer](https://micrometer.io) library.
You can expose the collected metrics to various monitoring systems, such as [Prometheus](https://prometheus.io) or [Dropwizard](https://metrics.dropwizard.io/)

>  for comprehensive monitoring and analysis.

Reviewer 3: ```suggestion
for comprehensive monitoring and analysis.
```

> To include the metrics service, simply add the service ID `metrics` to the `armeria-internal-services.include` configuration.

Reviewer: ```suggestion
To include the metrics service, simply add the service ID `metrics` to the `armeria.internal-services.include` configuration.
```

```java filename=ArmeriaConfiguration.java
@Bean
public MeterIdPrefixFunction meterIdPrefixFunction() {
    return MeterIdPrefixFunction.ofDefault("my.armeria.service");
}

@Bean
public MetricCollectingServiceConfigurator metricCollectingServiceConfigurator() {
    return metricCollectingServiceBuilder -> metricCollectingServiceBuilder
            .successFunction((context, log) -> {
                final int statusCode = log.responseHeaders().status().code();
                // Treat a 404 response as a success
                return statusCode >= 200 && statusCode < 400 || statusCode == 404;
            });
}
```

### Actuator support

If you are using [Spring Boot Actuator](https://docs.spring.io/spring-boot/docs/current/actuator-api/htmlsingle/),
you can serve it at the `armeria.internal-services.port`.
The internal services may also be served at the `management.server.port` if specified.

In order to use the actuator service, you need to first add the following dependencies.

<RequiredDependencies
  dependencies={[
    {
      groupId: 'com.linecorp.armeria',
      artifactId: 'armeria-spring-boot3-actuator-starter',
    }
  ]}
/>

> To include the actuator service, add the service ID `actuator` to the `armeria-internal-services.include` configuration.

Reviewer: ```suggestion
To include the actuator service, add the service ID `actuator` to the `armeria.internal-services.include` configuration.
```

# Using Spring Boot with Armeria

> Spring framework provides powerful features which are necessary for building a web application, such as

Reviewer 2: ```suggestion
Spring framework provides powerful features that are necessary for building a web application, such as
```

I heard that what comes after 'which' is something that's OK to omit completely. https://www.grammarly.com/blog/which-vs-that/

            return HttpResponse.of(HttpStatus.INTERNAL_SERVER_ERROR);
        }
        return HttpResponse.of(HttpStatus.CREATED);
    }
}
```

Inject this annotated service to the <type://ArmeriaServerConfigurator> bean.

```java filename=ArmeriaConfiguration.java
@Bean
public ArmeriaServerConfigurator armeriaServerConfigurator(TodoAnnotatedService todoAnnotatedService) {
   return serverBuilder -> {
       serverBuilder.serviceUnder("/docs", new DocService());
                    .decorator(LoggingService.newDecorator());
                    .annotatedService("/api", todoAnnotatedService);
   };
}
```

The following configuration prevents the execution of Spring's embedded web server and instead runs the Armeria.

```yaml filename=application.yml
# See https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto.webserver.disable.
# The application should not run as a web application and should not start an embedded web server.
spring.main.web-application-type: none

armeria:
  ports:
    - port: 8080
      protocols:
        - http
```

## Internal Services

Armeria supports four internal services which are helpful for monitoring and management.

> You can use those services by include the service ids to application.yml (or application.properties).

Reviewer 2: ```suggestion
You can use those services by include the service ids to `application.yml` (or `application.properties`).
```
(Please make all file names monospaced)

Include the service id `health` to `armeria-internal-services.include`. The `armeria.health-check-path` is not necessary.
The default path is `/internal/healthcheck`.

```yaml filename=application.yml
armeria:
  internal-services:
    include: health
  health-check-path: /internal/healthcheck
```

Make <type://HealthChecker> bean which implements your heath check business logic. For example, the below code is checking tomcat connector state.

```java filename=ArmeriaConfiguration.java
@Bean
public HealthChecker tomcatConnectorHealthChecker(ServletWebServerApplicationContext applicationContext) {
    final Connector connector = getConnector(applicationContext);
    return () -> connector.getState().isAvailable();
}
```

Similar with documentation service, you can add custom configuration of the service using <type://HealthCheckServiceConfigurator> bean.

```java filename=ArmeriaConfiguration.java
@Bean
public HealthCheckServiceConfigurator healthCheckServiceConfigurator() {
    return healthCheckServiceBuilder -> healthCheckServiceBuilder
            .updatable(true)
            .startUnhealthy();
}
```

> ### Metric Service

Reviewer 2: Please don't capitalize each word in the heading for consistency with other pages.

```suggestion
### Collecting metrics
```

  Otherwise, unexpected problems may occur.

</Tip>

You can add a prefix to id of collected metrics using <type://MeterIdPrefixFunction> bean.
Also, you can add custom configuration of the service using <type://MetricCollectingServiceConfigurator> bean.

```java filename=ArmeriaConfiguration.java
@Bean
public MeterIdPrefixFunction meterIdPrefixFunction() {
    return MeterIdPrefixFunction.ofDefault("my.armeria.service");
}

@Bean
public MetricCollectingServiceConfigurator metricCollectingServiceConfigurator() {
    return metricCollectingServiceBuilder -> metricCollectingServiceBuilder
            .successFunction((context, log) -> {
                final int statusCode = log.responseHeaders().status().code();
                // Treat a 404 response as a success
                return statusCode >= 200 && statusCode < 400 || statusCode == 404;
            });
}
```

For remind, you must not configure above settings using <type://ServerBuilder> directly through
<type://ArmeriaServerConfigurator> or `Consumer<ServerBuilder>`. It will be overrided by automatic configuration of Armeria.

> ### Actuator Service

Reviewer 2: ```suggestion
### Actuator support
```

If you are using [Spring Boot Actuator](https://docs.spring.io/spring-boot/docs/current/actuator-api/htmlsingle/),
you can serve this at `armeria.internal-services.port` and you can serve the internal services at `managment.server.port`.

In order to use the actuator service, you need to add the following dependencies first.

<RequiredDependencies
  dependencies={[
    {
      groupId: 'com.linecorp.armeria',
      artifactId: 'armeria-spring-boot3-actuator-starter',
    }
  ]}
/>

Include the service id `actuator` to `armeria-internal-services.include`.

```yaml application.yml
armeria:
  internal-services:
    include: actuator
```

> ## Other Bean Configuration

Reviewer 2: ```suggestion
## Other bean configurations
```

We will explain about some additional bean configurations that would be helpful to know.

### <type://ArmeriaServerConfigurator> vs `Consumer<ServerBuilder>`

You can use both <type://ArmeriaServerConfigurator> and `Consumer<ServerBuilder>` to configure the server using <type://ServerBuilder>.
Actually those two are same which are just function of `ServerBuilder -> void`. The most important difference is the order
in which they are applied to the server. Armeria first configures all <type://ArmeriaServerConfigurator> beans
and then applies all `Consumer<ServerBuilder>` beans.

If you have multiple <type://ArmeriaServerConfigurator>s or `Consumer<ServerBuilder>`s, you can set the order by
[@Order](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/annotation/Order.html) annotation.
One thing to note is that the default order of <type://ArmeriaServerConfigurator> is `0(zero)`
not the [`Ordered.LOWEST_PRECEDENCE`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/Ordered.html#LOWEST_PRECEDENCE).

### <type://DependencyInjector>

> Armeria can manually inject dependency using <type://DependencyInjector>. You can find the example at [1.17.0 release notes](/release-notes/1.17.0).

Reviewer 2: What do you think about explaining the default out of the box behavior first? e.g. by default, with spring boot integration enabled, Armeria uses Spring BeanFactory for injecting dependencies. You can change this behavior by ...

Author: Thanks! I agree with your opinion. However, It seems that what you mentioned is not the default behavior based on what I know. Armeria uses Spring BeanFactory for injecting dependencies when `armeria.enable-auto-injection=true`. Therefore, I explained dependency injection using Spring in the later part.

Make the <type://DependencyInjector> bean. It will replace the default dependency injector.
Therefore, this case, be aware of the instances that were automatically injected before may not be injected now.

```java filename=ArmeriaConfiguration.java
@Bean
public DependencyInjector dependencyInjector() {
    return DependencyInjector.ofSingletons(
            new BadRequestExceptionHandler(),
            new AuthDecorator((ctx, req) ->
                    CompletableFuture.supplyAsync(() -> req.headers().get(AUTHORIZATION).equals("auth-token"))
            )
    );
}
```

You can also use the dependency injector that uses [`BeanFactory`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/BeanFactory.html)
of Spring. Once you have configured the properties, you can easily create the beans to be injected just like when you use Spring.

```yaml filename=application.yml
armeria:
  enable-metrics: true

> ```

Reviewer 2: This should be placed somewhere else?
